### DISCUSSION TITLE
[DSL] A relationship between ... already exists, property to ignore this

### DISCUSSION THREAD

**Discussion by neopaf** [structurizr/java#288] [Q&A] [answered]
### Description

Two different people create two separate systems in two different git repos.

They both define that they communicate:
* Architect of A defines that A->B "action".
* Architect of B defines that A->B "action" (action is the same, obviously).

When I include both dsl files into my space to give readers an overview of all systems of my department, dear Structurizr complains of a (complete) duplicate.

Please consider a **property** to ignore these conflicts: say, `ignoreDuplicateRelationships`, default `false`.

Then each A and B architects could have errors in case they have local conflicts.
But for overall picture one could turn `ignoreDuplicateRelationships`=`true` and be happy?

### Priority

I have no budget and there's no rush, please add this feature for free

### More information

(thanks for AWESOME product!)

**Comment by simonbrowndotje** [comment]
This is working as designed, and I'd rather not start adding flags to disable various validation rules. You're welcome to add that functionality to your own fork though.

I'd recommend looking at the [enterprise examples](https://github.com/structurizr/examples/tree/main/enterprise) that show a better way to generate landscape view of a collection of systems.

**Reply by neopaf** [reply]
thanks for the pointer, Simon.
Alas, I can't find the landscape view here.
We wanted an overall schema, navigable down into all the systems and their internals.

All that stops us (as we see it) is extra validation, that just gets in the way of our happiness.

**Reply by simonbrowndotje** [reply]
> Alas, I can't find the landscape view here.
> We wanted an overall schema, navigable down into all the systems and their internals.

That's exactly what the demo does. Did you run it?

**Reply by neopaf** [reply]
I did, and I see separate systems, all full of internal life. Yet without links between `softwareSystem`s.
As I'm trying to communicate is that we have different architects of all these systems and each of them wants to show `softwareSystem`(his)<->`softwareSystem`(external) relationships.

Running [example1](https://github.com/structurizr/examples/tree/main/enterprise/src/main/resources/example1):
![Снимок экрана 2024-05-06 в 14 45 26](https://github.com/structurizr/java/assets/4362564/3d4d183e-3358-4bf2-ad37-f636535225f7)

**Reply by simonbrowndotje** [reply, marked as answer]
You need to run it using the instructions at https://github.com/structurizr/examples/tree/main/enterprise#example-1 rather than using Structurizr Lite.

```
./gradlew run -PmainClass=org.example.Example1
```

![image](https://github.com/structurizr/java/assets/1009874/1813cf4a-d229-4ca1-b470-71d1407cd660)

**Reply by neopaf** [reply]
Hmm. Thanks for pointing that out, Simon.
Alas, we're dreaming of something DSL only, maybe with a sprinkle of scripts....

Do you think it's possible to somehow script-create whatever `example1` code does? 
`enrichSystemLandscape` and `findAndCloneRelationships` in scripts?

**Reply by neopaf** [reply]
With these parts of the whole, where, say, this relationship
```
            webapp -> customerService "Manages customer data using" "JSON/HTTPS"
```
is mentioned only once -- we can `include` all the `.dsl` files into one master file.
And get all the relations displayed.

But then only the architect that created that relationship would see it -- others, in this example, `CustomerService` architect would not see it on diagrams that he/she creates, when he/she plays in local Structurizr instance.

**Reply by simonbrowndotje** [reply]
> Do you think it's possible to somehow script-create whatever example1 code does?
> enrichSystemLandscape and findAndCloneRelationships in scripts?

It should be, yes.

**Reply by neopaf** [reply]
I'm at a loss as to how to do that: how to come over the fact that `A->B` relationship exists in both pieces?
Looks like one needs to load spaces from workspace script rather than trying to include them into master-workspace .dsl, as I've tried originally.

Not sure how to do that from scripts. Any pointers, Simon?

**Reply by neopaf** [reply]
Currently I've come up with idea to disambiguate identical relationships by adding a `// suffix` to their description:
```
			SC -> CMD_ON.provQueue "Provisions //A"
			SC -> CMD_ON.provQueue "Provisions //B"
```

then this script `HideDuplicates.groovy` would find and drop the suffix, marking duplicates as it goes.

subsequently duplicates can be skipped in view like this
```
			exclude "relationship.tag==duplicate"
```

and strong relationships (those that have duplicates) also marked visually, `styles.dsl`
```
styles {
	relationship "strong" {
		thickness 10
	}
```

`HideDuplicates.groovy`:
```groovy
import com.structurizr.model.Relationship

def model = workspace.model
def relationships = model.relationships

Map<Tuple, Relationship> master = [:]

relationships.each {
	String desc = it.description?.replaceAll(/\s*\/\/.*/, '')
	if(desc != it.description) {
		Tuple key = new Tuple(it.source, it.destination, desc)
		Relationship existing = master[key]
		if(existing) {
			existing.description = desc
			existing.addTags "strong"
			it.addTags "duplicate"
		} else
			master[key] = it
	}
}
```


I'm not happy with it, but posting for the googlers benefit...

**Comment by neopaf** [comment]
> > Do you think it's possible to somehow script-create whatever example1 code does?
> > enrichSystemLandscape and findAndCloneRelationships in scripts?
> 
> It should be, yes.

Doesn't seem to be possible.
Structurizr Lite shows only 1 workspace at `http://localhost:8080/workspace`, even if I'd figure out how to add newly generated Workspace instance to some internal list (couldn't yet).

Maybe On-Premises could do that: execute script after loading something that would create/update some other workspace.
But this all seems cumbersome, and sub part relative to what you've suggested earlier.
